csharp/src/Google.Protobuf/JsonFormatter.cs

```diff
@@ -221,16 +221,27 @@ private bool WriteMessageFields(TextWriter writer, IMessage message, bool assume
             foreach (var field in fields.InFieldNumberOrder())
             {
                 var accessor = field.Accessor;
-                if (field.ContainingOneof != null && field.ContainingOneof.Accessor.GetCaseFieldDescriptor(message) != field)
+                // Populated lazily, so we don't end up boxing a load of values that
+                // we'll just discard due to field presence.
+                object value;
+                // If we know about presence for the field, use that to determine whether
+                // or not to write the value. We *don't* use this for (non-oneof) message fields,
+                // so that we can control whether null is written or not via FormatDefaultValues.
+                if (field.HasPresence && !(field.FieldType == FieldType.Message && field.ContainingOneof == null))
```

Comment:
Why can't this just be:

```
  if (field.HasPresence) {
```

Why exclude messages and oneofs from this? Any field with presence should be skipped if it is not present.

Reply:
It's for message fields that *aren't* in a oneof - so that if you've specified "format default values" you still get null. Otherwise this is a breaking change.